File: dictate/notify.py
"""Console-based notifier (no desktop notifications)."""

from dataclasses import dataclass


@dataclass
class Notifier:
    """Sends desktop notifications."""

    enabled: bool = True
    default_timeout_ms: int = 3000
    app_name: str = "Dictate Agent"

    def notify(
        self,
        title: str,
        message: str = "",
        icon: str = "dialog-information",
        timeout_ms: int = 0,
        replace: bool = True,
    ) -> None:
        """
        Send a desktop notification.

        Args:
            title: Notification title
            message: Notification body
            icon: Icon name
            timeout_ms: How long to show (0 = use default)
            replace: Replace previous notification with same tag
        """
        if not self.enabled:
            return

        # Instead of hitting notify-send/dunst, log to stdout so the daemon
        # still emits useful breadcrumbs when running interactively.
        line = f"[{self.app_name}] {title}"
        if message:
            line = f"{line}: {message}"
        print(line)

    def recording(self) -> None:
        """Show recording notification."""
        self.notify(
            "Recording...",
            "Toggle again to stop",
            "audio-input-microphone",
            30000,
        )

    def transcribing(self) -> None:
        """Show transcribing notification."""
        self.notify(
            "Transcribing...",
            "Processing speech",
            "emblem-synchronizing",
            30000,
        )

    def processing(self, model: str) -> None:
        """Show processing notification."""
        self.notify(
            f"Processing with {model}...",
            "Working on your request",
            "emblem-synchronizing",
            30000,
        )

    def done(self, text: str) -> None:
        """Show completion notification."""
        # Truncate long text
        display = text[:100] + "..." if len(text) > 100 else text
        self.notify(
            "Done!",
            display,
            "emblem-ok-symbolic",
            3000,
        )

    def error(self, message: str) -> None:
        """Show error notification."""
        self.notify(
            "Error",
            message[:100],
            "dialog-error",
            5000,
        )

    def no_speech(self) -> None:
        """Show no speech detected notification."""
        self.notify(
            "No speech detected",
            "Try speaking louder",
            "dialog-warning",
            2000,
        )

    def not_running(self) -> None:
        """Show daemon not running notification."""
        self.notify(
            "Dictate not running",
            "Start with: dictate-agent",
            "dialog-warning",
            3000,
        )


def check_notify_dependencies() -> list[tuple[str, str]]:
    """Notifications no longer depend on external tools."""
    return []

File: src/notify.rs
use notify_rust::{Notification, Timeout};
use tracing::error;

pub struct Notifier {
    enabled: bool,
    timeout_ms: u32,
    app_name: String,
}

impl Notifier {
    pub fn new(config: &crate::config::NotificationConfig) -> Self {
        Self {
            enabled: config.enabled,
            timeout_ms: config.timeout_ms,
            app_name: "Dictate Agent".into(),
        }
    }

    fn notify(&self, title: &str, message: &str, icon: &str, timeout_ms: Option<u32>) {
        if !self.enabled {
            return;
        }
        let timeout = timeout_ms.unwrap_or(self.timeout_ms);
        if let Err(e) = Notification::new()
            .appname(&self.app_name)
            .summary(title)
            .body(message)
            .icon(icon)
            .timeout(Timeout::Milliseconds(timeout))
            .show()
        {
            error!("Notification failed: {}", e);
        }
    }

    pub fn recording(&self) {
        self.notify("Recording...", "Speak now", "media-record", Some(30000));
    }

    pub fn transcribing(&self) {
        self.notify(
            "Transcribing...",
            "Processing audio",
            "media-playback-start",
            Some(30000),
        );
    }

    pub fn processing(&self, model: &str) {
        self.notify(
            "Processing...",
            &format!("Using {}", model),
            "system-run",
            Some(30000),
        );
    }

    pub fn done(&self, text: &str) {
        let display = if text.len() > 100 {
            &text[..100]
        } else {
            text
        };
        self.notify("Done", display, "dialog-ok", None);
    }

    pub fn error(&self, message: &str) {
        let display = if message.len() > 100 {
            &message[..100]
        } else {
            message
        };
        self.notify("Error", display, "dialog-error", Some(5000));
    }

    pub fn no_speech(&self) {
        self.notify(
            "No Speech",
            "No speech detected in recording",
            "dialog-warning",
            Some(2000),
        );
    }

    pub fn cancelled(&self) {
        self.notify(
            "Cancelled",
            "Recording discarded",
            "dialog-cancel",
            Some(2000),
        );
    }

    pub fn timer_set(&self, message: &str) {
        self.notify("Timer Set", message, "alarm-symbolic", None);
    }
}
